embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,905 @@
+//! be read out of the channel. In this case, all further attempts to send will
+//! result in an error.
+//!
+//! # Clean Shutdown
+//!
+//! When the [`Receiver`] is dropped, it is possible for unprocessed messages to
+//! remain in the channel. Instead, it is usually desirable to perform a "clean"
+//! shutdown. To do this, the receiver first calls `close`, which will prevent
+//! any further messages to be sent into the channel. Then, the receiver
+//! consumes the channel to completion, at which point the receiver can be
+//! dropped.
+//!
+//! This channel and its associated types were derived from https://docs.rs/tokio/0.1.22/tokio/sync/mpsc/fn.channel.html
+
+use core::cell::UnsafeCell;
+use core::fmt;
+use core::mem::MaybeUninit;
+use core::pin::Pin;
+use core::task::Context;
+use core::task::Poll;
+use core::task::Waker;
+
+use futures::Future;
+
+use super::CriticalSectionMutex;
+use super::Mutex;
+use super::NoopMutex;
+use super::ThreadModeMutex;
+
+/// Send values to the associated `Receiver`.
+///
+/// Instances are created by the [`split`](split) function.
+pub struct Sender<'ch, M, T, const N: usize>
+where
+    M: Mutex<Data = ()>,
+{
+    channel: &'ch Channel<M, T, N>,
+}
+
+// Safe to pass the sender around
+unsafe impl<'ch, M, T, const N: usize> Send for Sender<'ch, M, T, N> where M: Mutex<Data = ()> {}
+unsafe impl<'ch, M, T, const N: usize> Sync for Sender<'ch, M, T, N> where M: Mutex<Data = ()> {}
+
+/// Receive values from the associated `Sender`.
+///
+/// Instances are created by the [`split`](split) function.
+pub struct Receiver<'ch, M, T, const N: usize>
+where
+    M: Mutex<Data = ()>,
+{
+    channel: &'ch Channel<M, T, N>,
+}
+
+// Safe to pass the receiver around
+unsafe impl<'ch, M, T, const N: usize> Send for Receiver<'ch, M, T, N> where M: Mutex<Data = ()> {}
+unsafe impl<'ch, M, T, const N: usize> Sync for Receiver<'ch, M, T, N> where M: Mutex<Data = ()> {}
+
+/// Splits a bounded mpsc channel into a `Sender` and `Receiver`.
+///
+/// All data sent on `Sender` will become available on `Receiver` in the same
+/// order as it was sent.
+///
+/// The `Sender` can be cloned to `send` to the same channel from multiple code
+/// locations. Only one `Receiver` is valid.
+///
+/// If the `Receiver` is disconnected while trying to `send`, the `send` method
+/// will return a `SendError`. Similarly, if `Sender` is disconnected while
+/// trying to `recv`, the `recv` method will return a `RecvError`.
+///
+/// Note that when splitting the channel, the sender and receiver cannot outlive
+/// their channel. The following will therefore fail compilation:
+////
+/// ```compile_fail
+/// use embassy::util::mpsc;
+/// use embassy::util::mpsc::{Channel, WithThreadModeOnly};
+///
+/// let (sender, receiver) = {
+///    let mut channel = Channel::<WithThreadModeOnly, u32, 3>::with_thread_mode_only();
+///     mpsc::split(&channel)
+/// };
+/// ```
+pub fn split<M, T, const N: usize>(
+    channel: &Channel<M, T, N>,
```

Comment:
> Since this takes `&Channel`, it's actually possible to call it multiple times on the same channel. I'd argue it doesn't really "split" then, just "create a sender and receiver".
> 

It does actually split, although it should only be allowed to call it once. The question here for me is how can we achieve that?

Tokio and others do so by providing a `channel` function that creates the channel on the heap and returns the sender and a single receiver. We don't have the heap of course, and our channel buffer is created separately using a const generic size.

> One alternative is to require `&mut`, though I imagine this will impact usability.
> 

I'm not sure how this can be done.

> Another alteriative is to embrace it's possible to create senders/receivers on demand, and convert `split` into `sender` and `receiver` methods in Channel? In this case the channel would really be MPMC, maybe we should just call it "Channel" then, not specify the kind.

I was really going for MPSC from both: (a) the perspective of being able to optimise for that; and (b) being able to reason that a channel's value is consumed from just one place.

I'm not sure things are optimised for a single consumer now, but that is still my design goal.

Reply:
Other than advertising this an MPMC, I can't think of anything we can do to enforce the single consumer at compile time. Right now, I think the best we can do re. MPSC is to provide the runtime check as we do.

Reply:
It occurred to me that, by splitting a borrow on the channel, I can have the compiler enforce MPSC. I've now taken care of this as per commit 37710dba94d515465706d2700ac3b9ac5124a38e.

In essence, the sender and receiver structures reference the ChannelCell. Furthermore, for the receiver, it mutably borrows a `receiver_consumed` field, thus preventing another receiver existing for the same channel at the same time.